torch/utils/data/datapipes/iter/sharding.py

```diff
@@ -0,0 +1,77 @@
+from typing import (
+    Dict,
+    Sized,
+    Tuple,
+)
+
+from torch.utils.data.datapipes._decorator import functional_datapipe
+from torch.utils.data.datapipes.datapipe import IterDataPipe
+from torch.utils.data.datapipes.iter.grouping import SHARDING_PRIORITIES
+
+__all__ = [
+    "ShardingFilterIterDataPipe",
+]
+
+class _ShardingIterDataPipe(IterDataPipe):
+    def is_shardable(self):
+        return True
```

Comment:
TBH, I don't think this function is useful as all.

Reply:
Agree. So the API contract will be: 
* `_ShardingIterDataPipe` is always shardable (after all, the name is "Sharding DataPipe"
* It has "apply_sharding" method (in the future we may introduce other things like `is_replicable`.

Reply:
Yeah, exactly.

Reply:
> * Adding more built-in sharding DataPipe?

IDK.Probably yes when we have a solid use case

> * Users will want to use this to implement custom sharding?

 Yeah, I think so.

Reply:
Do we consider `[ShardingRoundRobinDispatcherIterDataPipe](https://github.com/pytorch/data/blob/01fc76200354501b057bb439b43a1f05f609dd0a/torchdata/datapipes/iter/util/sharding.py#L16-L17)` as a `_ShardingIterDataPipe` ? :)

Reply:
IMHO, this DataPipe should be labeled as `non-replicable` with `is_replicable` returning `False` and inherit `_ShardingIterDataPipe`.

In the `apply_sharding` function, we might want to only apply sharding to replicable `_ShardingIterDataPipe`

Reply:
Makes sense to me. Perhaps a good follow-up / bootcamp task to consolidate [ShardingRoundRobinDispatcherIterDataPipe](https://github.com/pytorch/data/blob/01fc76200354501b057bb439b43a1f05f609dd0a/torchdata/datapipes/iter/util/sharding.py#L16-L17) ?  (want to keep this PR as an initial step)

Reply:
SGTM.